doc/develop/coding-guidelines.rst

```diff
@@ -0,0 +1,213 @@
+******************************************
+
+This will help you to write code with the correct format. In general, we follow the
+guidelines stated in `PEP 8 <https://peps.python.org/pep-0008/>`_. Here we will list the most common practices used by a software
+developer for odemis.
+
+Naming Conventions
+==================
+
+* Class names should be in CamelCase in which multiple words are connected by capitalizing the first letter of the word. For e.g. MyClass().
+
+* Function name should be in snake_case in which the multiple words in small capitalization are joined with underscore where the first word is verb. For e.g. get_image()
+
+* Import modules should not be renamed. Use the original import name. For e.g. import numpy (and not import numpy as np)
+
+ * Exception: import matplotlib.pyplot as plt.
+
+* VAs in odemis should be in mixedCase which is different from PEP 8.
+
+
+Exception : If dependent on outer dependencies which follow different
+conventions. For instance, code inheriting from wxPython classes must follow
+the wxPython style, which follows the wxwidget style from
+typical C++ conventions.
+
+
+For e.g. 1.1 the VAs in odemis have the following structure
+
+.. code-block:: python
+
+ self.pixelSize = model.VigilantAttribute(pxs, unit="m", readonly=True)
+ self.horizontalFoV = model.FloatContinuous(hfv, range=[10e-9, 10e-3],
+                                               unit="m")
+ self.magnification = model.VigilantAttribute(self._hfw_nomag / hfv,
+                                                 unit="", readonly=True)
+
+* Naming of test files follow the below ways to provide consistency for files in test folders (no _init_.py file), which are included in each sub-directories of src folder.
+
+ * A unittest convention is used to name the test class for a module as TestModuleName(unittest.Testcase).
+
+ * Class methods are named as test_method_name().
+
+ * In a base class test which is used for repeating test cases with different input settings, we name the base class as TestModuleName.
+
+For e.g. 1.2 the naming of a base class test is as follows
+
+.. code-block:: python
+
+    class LakeshoreBaseTest:
+        """Base class for testing different models of LakeShore temperature controller."""
+
+        @classmethod
+        def tearDownClass(cls):
+            cls.dev.terminate()  # free up socket.
+
+In class, mark the attributes as private with a first letter as _. They are
+accessible normally from other classes, but by convention, as developers we know that they are
+private to the classes and should not be externally modified.
+If there are public attributes, the attributes might have to be changed in several
+files and folders.
+
+The usage of public attributes is useful when their access is required by a third
+party or another class. In this scenario, public attributes can be used.
+
+For e.g. 1.2 the implementation of public and private attributes
+
+.. code-block:: python
+
+    class Pokemon:
+        """
```

Comment:
```suggestion
        """Links special powers to the given pokemon type."""
```

For single-lined description, should `"""` be on the same line?

Reply:
I don't have strong feelings about it. We could say that both options are allowed. If we want to reduce alternatives, I'd suggest to go simple and say we should always have separate lines (this way, if extending the docstring after, there are less changes).